### PR TITLE
feat(helm): Allow configurable severity and thresholds for alerts

### PR DESCRIPTION
**Pull request by schahal**
**What this PR does / why we need it**:

The Loki helm chart comes with hardcoded `PrometheusRules` where users are unable to change the e.g., severity or threshold to best fit their environment's behavior.

This PR allows users to set some overrides.

For example, in my environment, a `critical` severity may mean a phone call in the middle of the night. So I have a values.yaml with the following override:

```
monitoring:
  rules:
    configs:
      LokiRequestPanics:
        # warn me if there are more than 3 panics over the last hour
        lookbackPeriod: 60m
        severity: warning
        threshold: 2
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

* This is simply pulling the default attributes of the PrometheusRule expressions into `values.yaml`
* This is _**backwards compatible**_ with the `monitoring.rules.disabled` keys (marking that as deprecated and to use configs for enabling/disabling). But we keep the deprecated keys for backwards compatibility
* **Testing**: Ran `helm template test --set loki.useTestSchema=true --set monitoring.rules.enabled=true --debug . --validate` and confirmed not rendered diff before and after.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

**Comment by schahal** [comment]
Hi @JStickler @DylanGuedes @chaudum @vlad-diachenko  : I'm wondering how we can get a reviewer/approval for this? I don't see any guidance in the Contributors doc or guidelines.

Tagged you as I saw the last couple `feat(helm)` approvals came from you (apologies in advance if it should be someone else, or if I missed something in `CONTRIBUTING.md`).

This is to address/resolve https://github.com/grafana/loki/issues/11219

**Comment by schahal** [comment]
Adding @trevorwhitney and @paul1r also for approval of the CI workflows and possible review/approve/merge?

**Comment by jkroepke** [comment]
@schahal Hello from new maintainer team! 🖐️

Do we still need this PR? If yes, please update it and add a CHANGELOG entry and remove the version bump. Thanks!

**Comment by schahal** [comment]
Hi @jkroepke , yeah I think this is still needed (as I've simply worked around it by disabling the rules altogether).

I've updated the branch to (a) remove the Chart version bump, (b) updated the CHANGELOG, (c) `additionalRuleAnnotations` has been added since I last worked on this branch, updated its rendering.

File: production/helm/loki/src/alerts.yaml.tpl [hunk 1]
```diff
@@ -2,96 +2,110 @@
...
-{{- end }}
+          severity: {{ .severity }}
+          {{- with $additionalLabels }}
+          {{ toYaml . | nindent 10 }}
+          {{- end }}
+      {{- end }}
 {{- end }}
 
-{{- if not (.Values.monitoring.rules.disabled.LokiCanaryLatency | default false) }}
+{{- if not (.disabled.LokiCanaryLatency | default false) }}
```
Comment (jkroepke) [inline production/helm/loki/src/alerts.yaml.tpl]: What about configs.LokiCanaryLatency.enabled here?

File: production/helm/loki/src/alerts.yaml.tpl [hunk 2]
```diff
@@ -2,96 +2,110 @@
 groups:
   - name: "loki_alerts"
     rules:
-{{- if not (.Values.monitoring.rules.disabled.LokiRequestErrors | default false) }}
+{{- with .Values.monitoring.rules }}
+{{- $additionalAnnotations := .additionalRuleAnnotations }}
+{{- $additionalLabels := .additionalRuleLabels }}
+{{- if not (.disabled.LokiRequestErrors | default (not .configs.LokiRequestErrors.enabled)) }}
```
Comment (jkroepke) [inline production/helm/loki/src/alerts.yaml.tpl]: I guess, this is an breaking change here. The problem here is that if `disabled.LokiRequestErrors` is set to false, then the evaluation should be set to true anyways, because `(not .configs.LokiRequestErrors.enabled)` resolve to true.

TL;DR: `false | default true` results into `true` as well.

File: production/helm/loki/src/alerts.yaml.tpl (hunk 1 above)
Comment (schahal) [reply production/helm/loki/src/alerts.yaml.tpl]: Whoops missed this during merge-conflict-resolutions (this branch was opened over a year ago and this was added since then).

Updated in latest PR.

File: production/helm/loki/src/alerts.yaml.tpl (hunk 2 above)
Comment (schahal) [reply production/helm/loki/src/alerts.yaml.tpl]: Updated in last commit. Good catch!!

`and()` is safer and cleaner to read in fact.

Tested by setting `disabled.LokiRequestErrors` to true and diff'd rendered before/after:

```diff
1515,1528d1514
<         - alert: LokiRequestErrors
<           annotations:
<             message: |
<               {{ $labels.job }} {{ $labels.route }} is experiencing {{ printf "%.2f" $value }}% errors.
<             shum: fa
<           expr: |
<             100 * sum(rate(loki_request_duration_seconds_count{status_code=~"5.."}[2m])) by (namespace, job, route)
<               /
<             sum(rate(loki_request_duration_seconds_count[2m])) by (namespace, job, route)
<               > 10
<           for: 15m
<           labels:
<             severity: critical
```

**Comment by jkroepke** [comment]
LGTM

**Comment by schahal** [comment]
Thanks @jkroepke . Are there any additional steps I need to take to get this merged?

**Review by Jayclifford345 — APPROVED** [review]
LGTM
